### PR TITLE
Support Multi-GPU training based on the paper "On Scaling Up 3D Gaussian Splatting Training" 

### PR DESCRIPTION
**Pull request by liruilong940607**
Paper link: https://daohanlu.github.io/scaling-up-3dgs/

Latest results:

Screenshot 2024-08-03 at 12 18 03 AM

Scripts: `bash benchmarks/basic_4gpus.sh`

### CODE REVIEW

**Comment by liruilong940607** [comment]
WIP on cleaning up

**Comment by ichsan2895** [comment]
It is strange that `Gsplat-MCMC` perform worse than `Gsplat-Default`.
I have shown some experiment in `discord` server. `Gsplat-MCMC` got better result despite it has lower GSs count. If it got worse, try change another hyperparameter such as `--opacity_reg 0.001` instead of the default `0.01`.

File: examples/simple_trainer.py [hunk 1]
```diff
@@ -180,11 +185,17 @@ def create_splats_with_optimizers(
...
-    N = points.shape[0]
     # Initialize the GS size to be the average dist of the 3 nearest neighbors
     dist2_avg = (knn(points, 4)[:, 1:] ** 2).mean(dim=-1)  # [N,]
     dist_avg = torch.sqrt(dist2_avg)
     scales = torch.log(dist_avg * init_scale).unsqueeze(-1).repeat(1, 3)  # [N, 3]
+
+    # Distribute the GSs to different ranks (also works for single rank)
+    points = points[world_rank::world_size]
+    rgbs = rgbs[world_rank::world_size]
+    scales = scales[world_rank::world_size]
```
Comment (liruilong940607) [inline examples/simple_trainer.py:196]: @TarzanZhao One question here. Any idea on the best way to split the GSs for multi-GPU initialization? I'm currently essentially randomly split them, which might not ideal for minimizing data transfer?

File: gsplat/rendering.py [hunk 2]
```diff
@@ -213,6 +232,30 @@ def rasterization(
...
+        torch.distributed.all_gather_object(C_world, C)
+
+        # [TODO] `all_gather` is not differentiable w.r.t. viewmats and Ks
+        out_tensor_list = [torch.empty((C_i, 4, 4), device=device) for C_i in C_world]
+        torch.distributed.all_gather(out_tensor_list, viewmats.contiguous())
+        viewmats = torch.cat(out_tensor_list, dim=0)
+        out_tensor_list = [torch.empty((C_i, 3, 3), device=device) for C_i in C_world]
+        torch.distributed.all_gather(out_tensor_list, Ks.contiguous())
+        Ks = torch.cat(out_tensor_list, dim=0)
+        C = len(viewmats)  # Silently change C from local #Cameras to global #Cameras.
```
Comment (liruilong940607) [inline gsplat/rendering.py:258]: @TarzanZhao  The current design of this API is that it takes in the local cameras that needs to be rendered by this rank. So before the projection stage, an all_gather needs to be done for each rank to get access to all cameras. 

I guess another option for the API design is that this function takes in the "global" cameras (all cameras needed to be rendered by all ranks), which would avoid this all_gather operation. But this would mean that user have to load the same global set of cameras in each rank, but only supervise on a subset of them, which feels a bit counter intuitive from the user experience perspective.

I personally prefer the first design choice but that means two all_gather and sync operations here. Do you think the affect of this would be large in your experience?

File: gsplat/rendering.py [hunk 3]
```diff
@@ -314,6 +354,137 @@ def rasterization(
         # make it apple-to-apple with Inria's CUDA Backend.
         colors = torch.clamp_min(colors + 0.5, 0.0)
 
+    # If in distributed mode, we need to scatter the GSs to the destination ranks, based
+    # on which cameras they are visible to, which we already figured out in the projection
+    # stage.
+    if distributed:
+        if packed:
```
Comment (liruilong940607) [inline gsplat/rendering.py:361]: @TarzanZhao This is the sparse all to all logic, which we called packed mode (only visible GSs are returned from the projection function).

I'm not sure if I'm implementing this part in a most efficient way. Would love to get your thoughts

**Comment by liruilong940607** [comment]
@TarzanZhao My biggest problem with this PR in its current stage is that it's far from getting 3.5x speedup on 4x GPUs. I'm not sure how much NVlinks play a role in this, which I don't have in my server setup. But would love to know if any thing can be improved on the implementation side, which I think its mostly just the update of the `rasterization()` function.

**Comment by liruilong940607** [comment]
I think I can live with like ~3x speedup on 4 GPUs because I'm not implementing the tile-based balancing and GS rebalancing logic. Is that a reasonable expectation?

**Comment by liruilong940607** [comment]
Using packed mode:

```bash
CUDA_VISIBLE_DEVICES=4,5,6,7 python simple_trainer.py --steps_scaler 0.25 --eval_steps -1 --disable_viewer --packed
>>> Step:  7499 {'mem': 2.944546699523926, 'ellipse_time': 1661.8717801570892, 'num_GS': 1356072}
>>> Step:  7499 {'mem': 3.056483268737793, 'ellipse_time': 1661.9940576553345, 'num_GS': 1432245}
>>> Step:  7499 {'mem': 3.076539993286133, 'ellipse_time': 1661.1862392425537, 'num_GS': 1430476}
>>> Step:  7499 {'mem': 3.133754253387451, 'ellipse_time': 1661.065690279007, 'num_GS': 1470615}
>>> PSNR: 27.431, SSIM: 0.8689, LPIPS: 0.075 Time: 0.087s/image Number of GS: 1356072
```

**Comment by liruilong940607** [comment]
As a reference. Running the [official repo](https://github.com/nyu-systems/Grendel-GS/tree/0ea84e456d58946aa9708e1932d7c3466edd6a98) in the same GPU environment (without NVlinks) get:

- 4 GPUs with total batch size 4:
```
# 4g_4b.sh
torchrun --standalone --nnodes=1 --nproc-per-node=4 train.py \
        -s ${dataset_folder}/${scene} \
        --images ${image_folder} \
        --llffhold 8 \
        --iterations 30000 \
        --log_interval 250 \
        --model_path ${expe_folder}/4g_4b/${expe_name} \
        --bsz $BSZ \
        $monitor_opts \
        --test_iterations 30000 \
        --save_iterations 30000 \
        --eval

Training Finished after 15m30s
[ITER 29997] Evaluating test: L1 0.028009627014398575 PSNR 27.24338150024414 [19/07 17:24:22]
[ITER 29997] Evaluating train: L1 0.02116350270807743 PSNR 29.852651596069336 [19/07 17:24:23]
```

File: gsplat/rendering.py (hunk 2 above)
Comment (TarzanZhao) [reply gsplat/rendering.py:258]: Using all_gather_object() is often slow in my experience because it transfers data from the CPU to the GPU, then to another GPU, and finally back to the CPU. It's advisable to avoid using this function as much as possible. If you prefer the first design, then please try to avoid using all_gather_object() at least. 

In our Grendel's implementation, every gpu keep the same dataset and uses the same random seed to generate the same batch across GPU every time. This is the other option that you have mentioned. Then there is no need for all-gather cameras.

File: gsplat/rendering.py (hunk 3 above)
Comment (TarzanZhao) [reply gsplat/rendering.py:361]: 1. depth and radii do not have to use functional version of all2all.
2. collected_cnts does not need functional version of all2all. 
3. [cnt.item() for cnt in cnts] will invoke many times of GPU-CPU communication. The total latency will be large. You can change to transfer cnts to cpu by single call. 

In general, this code has more kernels than ours: https://github.com/nyu-systems/Grendel-GS/blob/0ea84e456d58946aa9708e1932d7c3466edd6a98/gaussian_renderer/__init__.py#L176

More kernels will add significant kernel launch overheads, especially when images have small resolutions.

File: gsplat/rendering.py [hunk 4]
```diff
@@ -314,6 +354,137 @@ def rasterization(
...
+            ]
+            # differentiable all_to_all
+            DistF.all_to_all(collected_fl_list, data_fl_list)
+            collected_fl = torch.cat(collected_fl_list, dim=0)  # [C_i * N, :]
+            collected = collected_fl.reshape(
+                C, sum(N_world), data.shape[-1]
+            )  # [C_i, N, :]
+
+            # collected contains:
+            radii = collected[..., 0].int()
```
Comment (TarzanZhao) [inline gsplat/rendering.py:481]: You can use torch.split() to avoid many kernel launches here.

**Review by TarzanZhao — COMMENTED** [review]
Cool!! I'm very excited that our paper's distributed strategy can be adopted by gsplat so quickly. Thanks so much! I leave some comments and reference code and hope these can help a little bit.

**Comment by TarzanZhao** [comment]
> @TarzanZhao My biggest problem with this PR in its current stage is that it's far from getting 3.5x speedup on 4x GPUs. I'm not sure how much NVlinks play a role in this, which I don't have in my server setup. But would love to know if any thing can be improved on the implementation side, which I think its mostly just the update in the `rasterization()` function.

Overall, I think there are two major reasons: 
1. Data loading might be slow. 
3. There are more kernel launches than our implementation, but this is related to the data format required by gsplat. You want the code to be clear and modular, and for some functions to be decoupled, so some additional data format conversion kernels are needed. I think this is normal.

**Comment by TarzanZhao** [comment]
> PSNR

The PSNR for the garden scene in this test is lower than in our experiments (below) . Could you please run it again to check if this is due to numerical instability? 

image

**Comment by TarzanZhao** [comment]
> > PSNR
> 
> The PSNR for the garden scene in this test is lower than in our experiments (below) . Could you please run it again to check if this is due to numerical instability?
> 
> image



> As a reference. Running the [official repo](https://github.com/nyu-systems/Grendel-GS/tree/0ea84e456d58946aa9708e1932d7c3466edd6a98) in the same GPU environment (without NVlinks) get:
> 
> * 4 GPUs with total batch size 4:
> 
> ```
> # 4g_4b.sh
> torchrun --standalone --nnodes=1 --nproc-per-node=4 train.py \
>         -s ${dataset_folder}/${scene} \
>         --images ${image_folder} \
>         --llffhold 8 \
>         --iterations 30000 \
>         --log_interval 250 \
>         --model_path ${expe_folder}/4g_4b/${expe_name} \
>         --bsz $BSZ \
>         $monitor_opts \
>         --test_iterations 30000 \
>         --save_iterations 30000 \
>         --eval
> 
> Training Finished after 15m30s
> [ITER 29997] Evaluating test: L1 0.028009627014398575 PSNR 27.24338150024414 [19/07 17:24:22]
> [ITER 29997] Evaluating train: L1 0.02116350270807743 PSNR 29.852651596069336 [19/07 17:24:23]
> ```
> 
> Note the LPIPS is not the same LPIPS as in this repo.

The PSNR for the garden scene in this test is lower than in our experiments (below) . Could you please run it again to check if this is due to numerical instability?

image

**Comment by TarzanZhao** [comment]
> As a reference. Running the [official repo](https://github.com/nyu-systems/Grendel-GS/tree/0ea84e456d58946aa9708e1932d7c3466edd6a98) in the same GPU environment (without NVlinks) get:
> 
> * 4 GPUs with total batch size 4:
> 
> ```
> # 4g_4b.sh
> torchrun --standalone --nnodes=1 --nproc-per-node=4 train.py \
>         -s ${dataset_folder}/${scene} \
>         --images ${image_folder} \
>         --llffhold 8 \
>         --iterations 30000 \
>         --log_interval 250 \
>         --model_path ${expe_folder}/4g_4b/${expe_name} \
>         --bsz $BSZ \
>         $monitor_opts \
>         --test_iterations 30000 \
>         --save_iterations 30000 \
>         --eval
> 
> Training Finished after 15m30s
> [ITER 29997] Evaluating test: L1 0.028009627014398575 PSNR 27.24338150024414 [19/07 17:24:22]
> [ITER 29997] Evaluating train: L1 0.02116350270807743 PSNR 29.852651596069336 [19/07 17:24:23]
> ```
> 
> Note the LPIPS is not the same LPIPS as in this repo.

The PSNR for the garden scene in this test is lower than in our experiments (below) . Could you please run it again to check if this is due to numerical instability?

image

**Comment by liruilong940607** [comment]
After some optimization, the training time of 4 GPUs is brought down to 16m25s!

```bash
CUDA_VISIBLE_DEVICES=4,5,6,7 python simple_trainer.py --steps_scaler 0.25 --eval_steps -1 --disable_viewer --packed

>>> Step:  7499 {'mem': 2.8663759231567383, 'ellipse_time': 985.6521210670471, 'num_GS': 1352151}
>>> Step:  7499 {'mem': 2.971991539001465, 'ellipse_time': 984.3997764587402, 'num_GS': 1406792}
>>> Step:  7499 {'mem': 3.0060572624206543, 'ellipse_time': 985.4934339523315, 'num_GS': 1454676}
>>> Step:  7499 {'mem': 3.0007195472717285, 'ellipse_time': 985.5467941761017, 'num_GS': 1428084}
```

**Comment by liruilong940607** [comment]
MCMC on 4 GPUs

```bash
CUDA_VISIBLE_DEVICES=4,5,6,7 python simple_trainer_mcmc.py --steps_scaler 0.25 --eval_steps -1 --packed

>>> Step:  7499 {'mem': 2.4360504150390625, 'ellipse_time': 991.8759768009186, 'num_GS': 1000000}
>>> Step:  7499 {'mem': 2.403714179992676, 'ellipse_time': 992.4375550746918, 'num_GS': 1000000}
>>> Step:  7499 {'mem': 2.42301607131958, 'ellipse_time': 992.5172688961029, 'num_GS': 1000000}
>>> Step:  7499 {'mem': 2.411731243133545, 'ellipse_time': 991.4808518886566, 'num_GS': 1000000}

PSNR: 27.714, SSIM: 0.8744, LPIPS: 0.079 Time: 0.074s/image Number of GS: 1000000
```

**Comment by Ben-Mack** [comment]
This PR is great! I wonder that can it be used in a way to use 1 GPU to process multiple batch? 
So it will allow much larger scene to be trained with just 1 GPU! (larger scene in term number of gaussians, or num of images, ideally both)
I think this 1 GPU use case can help a lot of people to train high quality GS without highend GPU.
@liruilong940607 Do you think it is possible?

**Comment by liruilong940607** [comment]
@Ben-Mack  For a single GPU, I guess there isn't much you can do other than looping over images?
